Connector builder: Incremental sync documentation (#25238)
airbytehq/airbyte: docs/connector-development/connector-builder-ui/incremental-sync.md

> # Incremental sync

Reviewer: do we want to document the use of `stream_interval` anywhere?

Author: Good call, added an info box for that and updated the tutorial

This is especially important if there are a large number of records to sync and/or the API has tight request limits which makes a full sync of all records on a regular schedule too expensive or too slow.

Incremental syncs are usually implemented using a cursor value (like a timestamp) that delineates which data was pulled and which data is new. A very common cursor value is an `updated_at` timestamp. This cursor means that records whose `updated_at` value is less than or equal than that cursor value have been synced already, and that the next sync should only export records whose `updated_at` value is greater than the cursor value.

To use incremental syncs, the API endpoint needs to fullfil the following requirements:
* Records contain a date/time field that defines when this record was last updated (the "cursor field")
* It's possible to filter/request records by the cursor field

To learn more about how different modes of incremental syncs, check out the [Incremental Sync - Append](/understanding-airbyte/connections/incremental-append/) and [Incremental Sync - Deduped History](/understanding-airbyte/connections/incremental-deduped-history) pages.

## Configuration

> To configure incremental syncs for a stream in the connector builder, you have to specify how the records specify the **"last changed" / "updated at" timestamp**, the **initial time range** to fetch records for and **how to request records from a certain time range**.

Reviewer 2: ```suggestion
To configure incremental syncs for a stream in the connector builder, you have to specify how the records will represent the **"last changed" / "updated at" timestamp**, the **initial time range** to fetch records for and **how to request records from a certain time range**.
```

In the builder UI, these things are specified like this:
* The "Cursor field" is the property in the record that defines the date and time when the record got changed. It's used to decide which records are synced already and which records are "new"
* The "Datetime format" specifies the [format](https://docs.python.org/3/library/datetime.html#strftime-and-strptime-format-codes) the cursor field is using to specify date and time,

> * The "Cursor granularity" is the smallest time unit supported by the API to specify the time range to request records for

Reviewer: this should either have a link to some external docs or a description / enumeration of the values that are accepted. `PT1S` doesn't mean much without context

* The "Start datetime" is the initial start date of the time range to fetch records for. When doing incremental syncs, the second sync will overwrite this date with the last record that got synced so far. In most cases, it is defined by the end user when configuring a Source using your connector.

> * The "End datetime" is the end date of the time range to fetch records for. In most cases it's set to the current date and time when the sync is started to sync all changes that happened so far.

Reviewer: I think it's worth mentioning that `{{ now_utc() }}` is how the value is set to the current date and time

Author: I left that out because I want to make this the default setting in a simple dropdown as part of https://github.com/airbytehq/airbyte/issues/21927 (started with that today)

```
curl 'https://content.guardianapis.com/search?order-by=oldest&from-date=2023-04-09T00:00:00Z&to-date=2023-04-15T10:18:08Z'
```

The last encountered date will be saved as part of the connection - when the next sync is running, it picks up from the last record. Let's assume the last ecountered article looked like this:
```
{
  "id": "business/live/2023/apr/15/uk-bosses-more-optimistic-energy-prices-fall-ai-spending-boom-economics-business-live",
  "type": "liveblog",
  "sectionId": "business",
  "sectionName": "Business",
  "webPublicationDate": "2023-04-15T07:30:58Z",
}
```

Then when a sync is triggered for the same connection the next day, the following request is made:
```
curl 'https://content.guardianapis.com/search?order-by=oldest&from-date=2023-04-15T07:30:58Z&to-date=2023-04-16T10:18:08Z'
```

The `from-date` is set to the cutoff date of articles synced already and the `to-date` is set to the new "now".

## Advanced settings

The description above is sufficient for a lot of APIs. However there are some more subtle configurations which sometimes become relevant. 

### Step

> When incremental syncs are enabled, the connector is not fetching all records since the cutoff date at once - instead it's splitting up the time range between the cutoff date and the desired end date into intervals based on the "Step" configuration (by default it's set to 1 month)

Reviewer 2: > by default it's set to 1 month

How do we define "1 month"? Given that months are variable numbers of days and there are edge cases that may not have obvious defaults (e.g. what is 1 month from Jan 31?), I wonder if it would be preferable to choose another default (a week?) instead.

Author: We are using ISO 8601 durations (https://en.wikipedia.org/wiki/ISO_8601#Durations) so a month is a calendar month which can have varying length. I'm not sure about the perfect default, I think @girarda also put some thoughts into making this field optional and instead do a number-of-records based checkpointing which would be preferable as it makes it easier for the user.

However, AFAIK that's not planned for the immediate future  (before public beta) so I'm documenting the current state of affairs.

> For example if the "Step" is set to 10 days (`P10D`) for the Guardian articles stream described above and a longer time range, then the following requests will be performed:

Reviewer: Can we point to some docs or an explanation of what the values are? `P10D` is very cryptic

```

curl 'https://content.guardianapis.com/search?order-by=oldest&from-date=2023-01-01T00:00:00Z&to-date=2023-01-10T00:00:00Z'

curl 'https://content.guardianapis.com/search?order-by=oldest&from-date=2023-01-10T00:00:00Z&to-date=2023-01-20T00:00:00Z'

curl 'https://content.guardianapis.com/search?order-by=oldest&from-date=2023-01-20T00:00:00Z&to-date=2023-01-30T00:00:00Z'

...
```

After an interval is processed, the cursor value of the last record will be saved as part of the connection as the new cutoff date.

In most cases, the default step size is fine, there are two reasons to change it:

> * **The API is unreliable** and the cutoff date should be saved more often to prevent resync of a lot of records - if the "Step" size is a day, then at most one day worth of data needs to be resync in case the sync fails halfway through. However, a smaller step size might cause more requests to the API and more load on the system.

Reviewer 2: ```suggestion
* **The API is unreliable** we may want to reduce the step size. This results in more frequent saving of the "most recent" date, which prevents resync of a lot of records - if the "Step" size is a day, then at most one day worth of data needs to be resynced if the sync fails halfway through. However, it should be noted that a smaller step size might cause more requests to the API and more load on the system.
```

Author: I changed this paragraph based on @girarda s comments above

> * **The API is unreliable** and the cutoff date should be saved more often to prevent resync of a lot of records - if the "Step" size is a day, then at most one day worth of data needs to be resync in case the sync fails halfway through. However, a smaller step size might cause more requests to the API and more load on the system.

Reviewer: Can we focus on the checkpointing ability instead of blaming the API for being unreliable? This is also useful when Airbyte is responsible for the sync failure

* **The API requires the connector to fetch data in pre-specified chunks** - for example the [Exchange Rates API](https://exchangeratesapi.io/documentation/) makes the date to fetch data for part of the URL path and only allows to fetch data for a single day at a time

### Lookback window

The "Lookback window" specifies a duration that is subtracted from the last cutoff date before starting to sync.

> Same APIs update records over time but do not allow to filter or search by modification date, only by creation date. For example the API of The Guardian might change the title of an article after it got published, but the `webPublicationDate` still shows the original date the article got published initially.

Reviewer 2: ```suggestion
Some APIs update records over time but do not allow to filter or search by modification date, only by creation date. For example the API of The Guardian might change the title of an article after it got published, but the `webPublicationDate` still shows the original date the article got published initially.
```

In these cases, there are two options:

> * **Do not use incremental sync** and always sync the full set of records to always have a consistent state - depending on the amount of data this might not be feasable 

Reviewer: nit: this is also undesirable for some use cases even if the amount of data is small. This prevents the user from keeping a history of the changes (dedup + history)
